### PR TITLE
Provide the ability to disable port security


### PR DESCRIPTION
**Pull request by evrardjp**
For some fringe cases, the openstack network where the
instanciated caasp lies might have port security disabled,
or the deployer want to disable port security when creating
the network all together.

Without this patch, it is impossible to create a network
with port security disabled, and attaching security groups
to a port security disabled network leads to failures in
the OpenStack API.

This fixes the port security case by adding a boolean,
defaulting to true, to determine if port security is enabled,
and use its value when creating network or attaching security
groups.

### CODE REVIEW

**Comment by evrardjp** [comment]
Reversed the order of type and default to make lint happy.

**Comment by evrardjp** [comment]
Changed email address, and  configured boolean as string to pass linting. Type is removed as it is inferred from default value.

File: ci/infra/openstack/worker-instance.tf [hunk 1]
```diff
@@ -76,10 +76,10 @@ resource "openstack_compute_instance_v2" "worker" {
     name = "${var.internal_net}"
   }
 
-  security_groups = [
+  security_groups = "${var.enable_openstack_port_security == true ? [
```
Comment (bergmannf) [inline ci/infra/openstack/worker-instance.tf]: I think this is missing the closing `}`.
```suggestion
  security_groups = "${var.enable_openstack_port_security} == true ? [
```

File: ci/infra/openstack/master-instance.tf [hunk 2]
```diff
@@ -64,10 +64,10 @@ resource "openstack_compute_instance_v2" "master" {
     name = "${var.internal_net}"
   }
 
-  security_groups = [
+  security_groups = "${var.enable_openstack_port_security == true ? [
```
Comment (bergmannf) [inline ci/infra/openstack/master-instance.tf]: I think this is missing the closing `{`
```suggestion
  security_groups = "${var.enable_openstack_port_security} == true ? [
```

**Comment by evrardjp** [comment]
Pushed again, because whitespaces. Thanks make lint.

**Comment by evrardjp** [comment]
That won't work with the current packaging, would you be happy to bump the version required to at least 1.17 for the terraform-provider-openstack?
That would still keep terraform to .11 vs .12.

File: ci/infra/openstack/load-balancer.tf [hunk 3]
```diff
@@ -2,9 +2,7 @@ resource "openstack_lb_loadbalancer_v2" "lb" {
   name          = "${var.stack_name}-lb"
   vip_subnet_id = "${openstack_networking_subnet_v2.subnet.id}"
 
-  security_group_ids = [
-    "${openstack_compute_secgroup_v2.secgroup_master_lb.id}",
-  ]
+  security_group_ids = "${var.enable_openstack_port_security == true ? ["${openstack_compute_secgroup_v2.secgroup_master_lb.id}", ] : []}"
```
Comment (evrardjp) [inline ci/infra/openstack/load-balancer.tf]: This is wrong and I will fix it when I know we can bump terraform-provider-openstack

**Comment by qnetter** [comment]
I'm a little confused - does adopting this change require updating terraform and/or the terraform openstack provider? Or can it be applied to the versions we are already shipping?

**Comment by qnetter** [comment]
Also, @evrardjp - these questions via @mdarnell-suse...

Is the use case for this defect a deployment of CaaSP on top of OpenStack? 

Does this defect only show up if OpenStack port security groups are disabled? 

And is this defect particular to an ironic deployment of CaaSP or an instance-based CaaSP deployment?

**Comment by jsuchome** [comment]
I cannot answer the questions for @evrardjp , but at least something from my POV: This change is not doing anything to default skuba deployments, right? It's only adding an option for disabling port security, but user has to specifically choose it as terraform variable...

**Comment by evrardjp** [comment]
Sorry I am on holidays, so I missed this. Here are my answers

> Does adopting this change require updating terraform and/or the terraform openstack provider?

Yes. Terraform openstack provider needs to be bumped. There is still one version that's available for terraform 0.11 doing this kind of feature (I can double check which version when I am back from holidays). Of course, a bump of terraform to 0.12 would definitely help, as conditional code behaviour is greatly improved in that version.

> Or can it be applied to the versions we are already shipping?
It cannot, as the openstack provider is ... ancient.

> Also, @evrardjp - these questions via @mdarnell-suse...
> 
> Is the use case for this defect a deployment of CaaSP on top of OpenStack?

It's to cover a specific use case, when someone wants to create virtual ips, virtual nics in openstack.
Port security prevents this, and therefore needs to be disabled. Instead of doing so manually on each port after the fact, we should IMO support it from the start by adding a few conditionals.

> 
> Does this defect only show up if OpenStack port security groups are disabled?

Well, if port security groups are enabled, you can't have your own vip around, having the ability to disable port security on the network would basically allow any user workload.

> 
> And is this defect particular to an ironic deployment of CaaSP or an instance-based CaaSP deployment?

That could match both, assuming the ironic deployed one is using neutron to create the network (which is not mandatory, afaik).

**Comment by toabctl** [comment]
@evrardjp @Itxaka needs rebase...

**Comment by evrardjp** [comment]
> @evrardjp @Itxaka needs rebase...

Needs work in general, as this won't work with the current state -- We just need to clarify supported versions of terraform and its openstack provider.

**Comment by ereslibre** [comment]
This PR has some conflicts that need resolving.

**Comment by evrardjp** [comment]
I will refresh this based on 0.12, when 0.12 support lands.

**Comment by evrardjp** [comment]
We should not have this until terraform is moved to 0.12.

**Comment by jordimassaguerpla** [comment]
Hi!
New code has been merged in master regarding terraform 0.12. Please rebase your branch to take this changes in.
Happy hacking

**Comment by evrardjp** [comment]
Oh wow that's ancient. I suppose not many people want to deploy things that rely on specific network stack on top of kubernetes on top of openstack...

**Pull request closed without merging**